### DISCUSSION TITLE
Can someone explain the state_action parameter to me?

### DISCUSSION THREAD

**Discussion by Sciamano72** [Koenkk/zigbee2mqtt#18312] [General]
Hello, I have tried searching everywhere for this but couldn't find any information.
Every ZigBee device I have installed has the following option in the "Settings (specific)" tab:

`state_action: State actions will also be published as 'action' when true (default false). The value must be true or false`

Could someone please explain what's this for? I really don't understand what it means/does and I never know whether I should select true or false.

Thank you very much.

**Comment by Sciamano72** [comment]
No one? :(

**Comment by fortiko** [comment]
I just came here to investigate the same mystery. Whenever we find out what that setting actually does, we should push to update the documentation to something clearer.

**Reply by fortiko** [reply]
@Sciamano72 The following PR added the feature "state_action" in the beginning of 2023: 
https://github.com/Koenkk/zigbee-herdsman-converters/pull/5276

From my understanding, the setting works like this:

If state_action is true, a [button] switch will, when executing its default action, additionally publish a message to the Action topic. The button press mimics a momentary action so that other listeners/automations can be triggered.

Example: Light A is physically controlled by a switch S. If switch S has state_action set to true, every switch/button press will publish a message to the Action topic, **in addition to controlling the light A itself**. You could now subscribe to topic Action and listen for those messages generated by switch S. Those messages could then trigger a second light B or start a more complex automation.

**Reply by Sciamano72** [reply]
Great find, this makes things much clearer.

**Reply by fortiko** [reply]
We could create a PR to update the documentation. I don't have access to physical buttons for the next several days, so if you could test that theory…?

**Reply by Sciamano72** [reply]
Unfortunately I can't, either. I actually have found the state_action option in a couple of switches which I'm not using with lights, though.

**Reply by fortiko** [reply]
Yes, the "default action" for different types of switches and buttons can vary. Some are light switches, others could be buttons on a smart relay for instance. That's why I mention "light" only in the example, not in the actual description of the feature:

"If state_action is true, a [button] switch will, when executing its default action, additionally publish a message to the Action topic. The button press mimics a momentary action so that other listeners/automations can be triggered."

**Reply by Sciamano72** [reply]
Yes, I am aware of that. What I mean is that I use the switches that expose the state_action option in such a way that I can't access their physical buttons. If I manage to get my hands on another one I will definitely test this theory :)

**Reply by fortiko** [reply]
Perhaps the physical button is not actually necessary, I was just not considering alternatives. A Zigbee smart plug can be triggered remotely of course in addition to a button press. Both "state" changes should lead to a message in Action, if `state_action` is enabled. From what I understand, this is something you could already observe with your current setup, correct?

**Reply by Sciamano72** [reply]
Er... I guess so. I'm not much of an HA expert, I've just started a few months ago, and many features are still mysterious to me. The Action thing is one of these: I get the concept but I actually don't really know how to use or even take advantage of it. I still have to learn a lot

**Comment by Charmio1** [comment]
I'm also struggling to find any decent explanation so I resorted to asking ChatGPT 4o, it's given by far the best explanation I've found (assuming it's accurate!). I figured I may as well share it:

**My prompt:** My fingerbot has a setting i dont understand in Z2M: "state_action: State actions will also be published as 'action' when true". Can you please explain?

**Responce:**
The state_action setting in your Fingerbot configuration relates to how the device reports actions back to Home Assistant. Here’s what it means:

When state_action is set to true: Every time the Fingerbot performs an action (like pressing a button), it will also publish an “action” event. This event is essentially a message that Home Assistant can pick up, letting you know the Fingerbot’s current state or action.

For example, if the Fingerbot presses the button, it will not only change the state (e.g., from off to on) but will also publish an “action” event to the MQTT topic (if you’re using MQTT). This can be useful if you want to create automations that specifically listen for these action events or track when the Fingerbot does something.


When state_action is set to false: The Fingerbot will only publish its “on” or “off” state without publishing additional “action” events. This mode is simpler but doesn’t allow for tracking individual actions as separate events.


How It Helps in Automations

If you enable state_action, you can use the “action” event to trigger specific automations. For instance, if you want to confirm each button press by capturing the action event in Home Assistant, setting state_action to true would be helpful. You’d be able to create automations that react every time the Fingerbot reports an action, making the setup more responsive and potentially more reliable.

**Comment by jamesonuk** [comment]
AIUI this will send some additional MQTT messages which will allow for automations.
Eg. on the Ikea bulbs I have this option exists,   With it set to false there is no action against the device. but if you set it to true you will see action in the state
![image](https://github.com/user-attachments/assets/46eb08e2-b685-4aea-83f1-88c79087d7e8)
and in MQTT you will start to see things like
![image](https://github.com/user-attachments/assets/e7a8b79b-07ae-47e6-a906-68d590e38d60)

then in HA automations you can use device automations like
![image](https://github.com/user-attachments/assets/3103097c-a58c-4199-9026-f7e1c1bdea6b)
You already have turned on / off automations (which I think are provided by the light platform).   I guess that the other on/off actions are based on the state in HA rather than MQTT so potentially marginally quicker or more history?

Not sure this particularly gives you much to go on and things like changing the brightness aren't actions so don't get logged but you can then automate things when lights are turned on (which you can already do)

Not sure I have ever found a reason to enable it but I am sure there are uses for it (I guess if it is say a device that doesn't have the actions built in??)
